ruby/examples/lib/apply_order_discount.rb

```diff
@@ -0,0 +1,44 @@
+class ApplyOrderDiscount
+  Request = Struct.new(:id)
+  Response = Struct.new(:discount, :has_discount?)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def apply(request)
+    order = gateway.find_order_by_id(request.id)
+
+    unless order.has_discount?
+      order.discount = discount_amount
+      gateway.save_order_discount(request.id, order)
+    end
+
+    Response.new(order.discount, order.has_discount?)
+  end
+
+  private
+
+  def discount_amount
+    10
+  end
+
+  class DiscountableOrder < Struct.new(:discount)
```

Comment:
In java you'd create interfaces to chop up the domain object, to hide aspects you don't care about.

Or even

class DiscountableOrder < Order 

Usecases requiring Order think it's an order 

If it quacks like a duck, swims like a duck and looks like a duck... well this we've got a duck